### PR TITLE
Migrate to cobra

### LINKED ISSUE #82: Running 'coder urls' without any arguments errors
```
> coder urls
2020-07-30 10:30:44 INFO	found ["work" "nathan" "hello" "commit-judge" "mydev" "nix"]
2020-07-30 10:30:44 FATAL	environment "" not found
```

We should instead check that the environment name isn't empty. If it is, we should print usage.

### LINKED ISSUE #85: Migrate to urfave/cli
Although https://github.com/cdr/cli offers a minimal wrapper, it's becoming clear that it lacks sufficient features for the growing scope of `coder-cli`. I propose we migrate to ~https://github.com/spf13/cobra~ https://github.com/urfave/cli. Consider a few advantages that we'll gain "out of the box". 

- Suggestions when "unknown command" happens
- bash, zsh, fish, and powershell completion generation
- well designed and standard command validation abstractions
- better persistent flag abstractions
- no more interface upgrading nonsense
- the model of a separate type per command is flawed and, given the lack of generics, is very difficult to abstract

Any thoughts / objections? cc @scsmithr @nhooyr @ammario

### PR DESCRIPTION
**Pull request by cmoog**
~I plan on following this up immediately with a cleanup of the error handling across the CLI layer. I'd prefer to merge as soon as posable to prevent blocking additional work. We should move away from calling `flog.Fatal` and towards the built-in error handling present in the `urfave/cli`.~ Done.  

As noted by @nhooyr, the closure pattern is non-obvious. Let me know what you think.

This PR also includes many improvements/additions, most notably to the `envs` and `urls` subcommands. `urls` was kept at the top level to prevent unnecessary coupling, but I'm thinking we should move it to a subcommand of `envs`. 
```
coder provides a CLI for working with an existing Coder Enterprise installation

Usage:
  coder [command]

Available Commands:
  completion  Generate completion script
  config-ssh  Configure SSH to access Coder environments
  envs        Interact with Coder environments
  help        Help about any command
  login       Authenticate this client for future operations
  logout      Remove local authentication credentials if any exist
  secrets     Interact with Coder Secrets
  sh          Open a shell and execute commands in a Coder environment
  sync        Establish a one way directory sync to a Coder environment
  urls        Interact with environment DevURLs
  users       Interact with Coder user accounts

Flags:
  -h, --help      help for coder
  -v, --version   version for coder

Use "coder [command] --help" for more information about a command.
```

Closes #85 
cc @nhooyr @anthonyshull @coadler 

Closes #82

### CODE REVIEW

File: cmd/coder/secrets.go [hunk 1]
```diff
@@ -7,60 +7,131 @@ import (
...
+				Usage:     "View a secret by name",
+				ArgsUsage: "[secret_name]",
+				Action:    viewSecret,
+			},
+		},
 	}
 }
 
-type listSecretsCmd struct{}
+func makeCreateSecret() cli.Command {
```
Comment (nhooyr) [inline cmd/coder/secrets.go:45]: Idk how I feel about this.

File: cmd/coder/envs.go [hunk 2]
```diff
@@ -1,29 +1,61 @@
...
+		Name:        "envs",
+		Usage:       "Interact with Coder environments",
+		Description: "Perform operations on the Coder environments owned by the active user.",
+		Action:      exitHelp,
+		Subcommands: []cli.Command{
+			{
+				Name:        "ls",
+				Usage:       "list all environments owned by the active user",
+				Description: "List all Coder environments owned by the active user.",
+				ArgsUsage:   "[...flags]>",
```
Comment (coadler) [inline cmd/coder/envs.go:24]: Is the trailing `>` intentional?

File: cmd/coder/envs.go [hunk 3]
```diff
@@ -1,29 +1,61 @@
...
+						return err
+					}
+
+					switch outputFmt {
+					case "human":
+						err := xtabwriter.WriteTable(len(envs), func(i int) interface{} {
+							return envs[i]
+						})
+						if err != nil {
+							return xerrors.Errorf("failed to write table: %w", err)
```
Comment (coadler) [inline cmd/coder/envs.go:38]:
```suggestion
							return xerrors.Errorf("write table: %w", err)
```

File: cmd/coder/envs.go [hunk 4]
```diff
@@ -1,29 +1,61 @@
...
+						}
+					case "json":
+						err := json.NewEncoder(os.Stdout).Encode(envs)
+						if err != nil {
+							return xerrors.Errorf("failed to write environments as JSON: %w", err)
```
Comment (coadler) [inline cmd/coder/envs.go:43]:
```suggestion
							return xerrors.Errorf("write environments as JSON: %w", err)
```

File: go.mod [hunk 5]
```diff
@@ -13,8 +13,8 @@ require (
 	github.com/mattn/go-colorable v0.1.6 // indirect
 	github.com/pkg/browser v0.0.0-20180916011732-0a3d74bf9ce4
 	github.com/rjeczalik/notify v0.9.2
-	github.com/spf13/pflag v1.0.5
-	go.coder.com/cli v0.5.0
+	github.com/stretchr/testify v1.6.1 // indirect
+	github.com/urfave/cli v1.22.4
```
Comment (coadler) [inline go.mod:17]: Why aren't we using `v2`?

File: cmd/coder/configssh.go [hunk 6]
```diff
@@ -11,162 +11,168 @@ import (
...
+	return func(c *cli.Context) error {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		currentConfig, err := readStr(*filepath)
+		if os.IsNotExist(err) {
+			// SSH configs are not always already there.
+			currentConfig = ""
+		} else if err != nil {
+			return xerrors.Errorf("failed to read ssh config file %q: %w", filepath, err)
```
Comment (coadler) [inline cmd/coder/configssh.go:69]: Should just ctrl f and fix these bad boys

**Comment by nhooyr** [comment]
Do you prefer v2 or v1 urfave @cmoog?

**Comment by coadler** [comment]
From looking that the migration docs it just seems like they made envs accept a slice and made most of the command/flag structs pointers instead.

**Comment by cmoog** [comment]
It looks like they enforce that flags are positioned before arguments too. Definetly not a fan of that. For example, the following is invalid.
```
coder secrets create newsecret --from-literal 123
```

In v2, this must be 
```
coder secrets create --from-literal 123 newsecret
```

**Comment by nhooyr** [comment]
Yea that straight sucks :(

**Comment by cmoog** [comment]
Yeah for sure, going to revert the migration cause that's super annoying.

**Comment by nhooyr** [comment]
I’d be in favour of using cobra then.

**Comment by cmoog** [comment]
Nevermind, I think the flag-before-arg is fine given that the `help` template accurately reflects that and we'll show plenty of example usages in the documentation.

**Comment by nhooyr** [comment]
From a UX perspective it's super annoying imo when I'm editing a CLI command. Any context on why they made this change?

**Comment by coadler** [comment]
@nhooyr https://github.com/urfave/cli/blob/master/docs/migrate-v1-to-v2.md#flags-before-args

> In v2 flags must come before args. This is more POSIX-compliant.

**Comment by nhooyr** [comment]
That’s so dumb. What does it have to do with POSIX lol.

**Comment by nhooyr** [comment]
https://github.com/urfave/cli/issues/1113

**Comment by cmoog** [comment]
@nhooyr @coadler how do you cast your votes?

**Comment by nhooyr** [comment]
If you're ok migrating, I'd say we go for cobra instead for UX. But I defer to you.

**Comment by cmoog** [comment]
There are some error handling inconsistencies but nothing too bad that can't be fixed after this merges. Don't want to keep this unmerged much longer as it's blocking feature work.

File: ci/README.md [hunk 7]
```diff
@@ -0,0 +1,26 @@
+# ci
```
[the remark this answers is not in the document]
Comment (cmoog) [reply ci/README.md:1]: Most readers of the `README` will be customers, not developers. Doesn't really make sense to bring attention to it in my view.

File: ci/integration/users_test.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+func TestUsers(t *testing.T) {
+	t.Parallel()
+	ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+	defer cancel()
+
+	c, err := tcli.NewContainerRunner(ctx, &tcli.ContainerConfig{
+		Image: "codercom/enterprise-dev",
+		Name:  "users-cli-tests",
+		BindMounts: map[string]string{
+			binpath: "/bin/coder",
```
[the remark this answers is not in the document]
Comment (cmoog) [reply ci/integration/users_test.go:22]: Doesn't that still require package level global state though?

File: ci/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (nhooyr) [reply ci/README.md:1]: Don't have to bring attention to it. Just put a link at the bottom. Customers/potential customers who care to look closely will be happy we document and care about these things.

File: ci/README.md [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+Each test group can have its own container for executing commands in a consistent
+and isolated filesystem.
+
+### prerequisites
+
+Assign the following environment variables to run the integration tests
+against an existing Enterprise deployment instance.
+
+```bash
+export CODER_URL=...
```
[the remark this answers is not in the document]
Comment (cmoog) [reply ci/README.md:17]: https://github.com/cdr/enterprise/issues/4702 I'm thinking that this means we'll be moving away from `CE`.

File: ci/integration/users_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (nhooyr) [reply ci/integration/users_test.go:22]: Yes but it's clear how things work. You can just jump to the definition to see how the function works whereas with `binpath`, it seems more magical to me. In general I think it's better to have a method that does something lazily and caches it than use an init function with a variable. It's not a big deal though, if you don't think it's an improvement, feel free to leave as is.

[the remark this answers is not in the document]
Comment (cmoog) [reply ci/integration/users_test.go:22]: added a comment

**Comment by nhooyr** [comment]
I'm a little swamped right now, don't think I can give this a good review sorry!

**Comment by cmoog** [comment]
To unblock feature work and prevent a rush before the release date, I'm going to merge this now. Some cleanup PRs will be following in parallel.

**Pull request merged**